### PR TITLE
CircleCI Secure Joining

### LINKED ISSUE #17060: Machine ID: Support secure joining on CircleCI
Reference GitHub implementation at https://github.com/gravitational/teleport/pull/16938

https://circleci.com/blog/openid-connect-identity-tokens/

https://circleci.com/docs/openid-connect-tokens#format-of-the-openid-connect-id-token

Down the road, we will want to create an orb: https://circleci.com/docs/orb-author

For the JWT to be injected, at least one context must be assigned to the workflow, yielding a token with the following claims:

```json
{
  "kid": "aK9j6dYBo3KDw4gfWy8xdMLU5Q8OycXh3sqn8J2h4NtEDxCDVyIJ_REDI8ODJjbUocTFJftRZ9UnLjzr9F1BFg",
  "typ": "JWT",
  "alg": "RS256"
},
{
  "sub": "org/00f45056-5918-4171-9222-b538c76ed3f8/project/dabad1c2-62eb-48a6-87ab-fc6ae6589281/user/688c029b-dfed-45bc-af7c-20281a884e2d",
  "aud": "00f45056-5918-4171-9222-b538c76ed3f8",
  "oidc.circleci.com/project-id": "dabad1c2-62eb-48a6-87ab-fc6ae6589281",
  "oidc.circleci.com/context-ids": [
    "05ad3418-d221-484f-99fb-46dc6b0f569b"
  ],
  "iss": "https://oidc.circleci.com/org/00f45056-5918-4171-9222-b538c76ed3f8",
  "exp": 1665056032,
  "iat": 1665052432
}
```

I recommend we allow the user to configure their org id, project id and also optionally specify a context id they want to allow.

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/17060

Implements support for CircleCI secure joining using OIDC joining with a similar implementation to GitHub secure joining.

## Sample Configs

```yaml
kind: token
version: v2
metadata:
  name: circleci-bot
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: [Bot]
  join_method: circleci
  bot_name: $BOT_NAME
  circleci:
    organization_id: $ORG_ID
    allow:
      - project_id: $PROJECT_ID
        context_id: $CONTEXT_ID
```

These IDs must be the UUIDs that you can find within CircleCI, and not the actual names of the org, project or context.

```yaml
# Use the latest 2.1 version of CircleCI pipeline process engine.
# See: https://circleci.com/docs/2.0/configuration-reference
version: 2.1

# Define a job to be invoked later in a workflow.
# See: https://circleci.com/docs/2.0/configuration-reference/#jobs
jobs:
  teleport:
    # Specify the execution environment. You can specify an image from Dockerhub or use one of our Convenience Images from CircleCI's Developer Hub.
    # See: https://circleci.com/docs/2.0/configuration-reference/#docker-machine-macos-windows-executor
    docker:
      - image: cimg/base:stable
    # Add steps to the job
    # See: https://circleci.com/docs/2.0/configuration-reference/#steps
    steps:
      - checkout
      - run:
          name: "teleport"
          command: |
            ./tbot start --auth-server=the.proxy:443 --join-method=circleci --token=circleci-bot --oneshot --destination-dir=./certs --data-dir=./data
            ./tsh -i ./certs/identity --proxy the.proxy:443 ssh noahstride@noah-laptop "echo $CIRCLE_BUILD_URL >> ~/circle_run_log"

# Invoke jobs via workflows
# See: https://circleci.com/docs/2.0/configuration-reference/#workflows
workflows:
  teleport-workflow:
    jobs:
      - teleport:
          context:
            - test-context
```

Notes:
- Context must be configured in CircleCI job or token will not be injected into the environment.

### CODE REVIEW

**Comment by zmb3** [comment]
Is `--join-method` really necessary if we can derive the join method from the specified token?

**Comment by strideynet** [comment]
> Is `--join-method` really necessary if we can derive the join method from the specified token?

The issue is that `tbot` won't be able to fetch the token to determine that until it has registered, unless we add an unauthenticated endpoint for expressing this information.

File: api/types/provisioning.go [hunk 1]
```diff
@@ -410,3 +426,23 @@ func (a *ProvisionTokenSpecV2GitHub) checkAndSetDefaults() error {
...
+	if len(a.Allow) == 0 {
+		return trace.BadParameter("the %q join method requires at least one token allow rule", JoinMethodCircleCI)
+	}
+	if a.OrganizationID == "" {
+		return trace.BadParameter("the %q join method requires 'organization_id' to be set", JoinMethodCircleCI)
+	}
+	for _, rule := range a.Allow {
+		projectSet := rule.ProjectID != ""
+		contextSet := rule.ContextID != ""
+		if !(projectSet || contextSet) {
```
Comment (zmb3) [inline api/types/provisioning.go:440]: Are project and context both required? The logic here says yes, but the error message makes it sound like only one is required.

File: lib/auth/join_circleci.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/gravitational/teleport/api/types"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/circleci"
+	"github.com/gravitational/trace"
+)
+
+func (a *Server) checkCircleCIJoinRequest(ctx context.Context, req *types.RegisterUsingTokenRequest) error {
+	if req.IDToken == "" {
+		return trace.BadParameter("%q not provided for %q join request", "IDToken", types.JoinMethodCircleCI)
```
Comment (zmb3) [inline lib/auth/join_circleci.go:30]: any reason not to just do?

```suggestion
		return trace.BadParameter("IDToken not provided for %q join request",  types.JoinMethodCircleCI)
```

File: lib/auth/join_circleci.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	pt, err := a.GetToken(ctx, req.Token)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	token, ok := pt.(*types.ProvisionTokenV2)
+	if !ok {
+		return trace.BadParameter("%q join method only support ProvisionTokenV2", types.JoinMethodCircleCI)
```
Comment (zmb3) [inline lib/auth/join_circleci.go:38]:
```suggestion
		return trace.BadParameter("%q join method only supports ProvisionTokenV2", types.JoinMethodCircleCI)
```

What do you think about including the actual type in the error message? Too much info?

File: lib/circleci/token_validator_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	require.NoError(t, err)
+
+	signer, err := jose.NewSigner(
+		jose.SigningKey{Algorithm: jose.RS256, Key: privateKey},
+		(&jose.SignerOptions{}).WithType("JWT"),
+	)
+	require.NoError(t, err)
+
+	providerMux := http.NewServeMux()
+	srv := httptest.NewServer(providerMux)
```
Comment (zmb3) [inline lib/circleci/token_validator_test.go:91]:
```suggestion
	srv := httptest.NewServer(providerMux)
	t.Cleanup(src.Close)
```

File: lib/githubactions/token_validator_test.go [hunk 5]
```diff
@@ -43,8 +43,8 @@ func fakeGithubIDP(t *testing.T) (*httptest.Server, jose.Signer) {
 	)
 	require.NoError(t, err)
 
-	url := ""
 	providerMux := http.NewServeMux()
+	srv := httptest.NewServer(providerMux)
```
Comment (zmb3) [inline lib/githubactions/token_validator_test.go:47]: Same as above, don't forget to shut down the server.

**Review by zmb3 — APPROVED** [review]

File: lib/auth/join_circleci.go (hunk 3 above)
Comment (strideynet) [reply lib/auth/join_circleci.go:38]: I can't see the harm in including it. As this error really will only ever be the result of programmer error, I can actually see it as being more helpful to include the type that has managed to sneak in. I'll add this.

File: api/types/provisioning.go (hunk 1 above)
Comment (strideynet) [reply api/types/provisioning.go:440]: Hmm, I'm pretty sure this behaves as the error describes (as in, only a single one must be set for it to be valid). I could rewrite it as:
```
if (!projectSet && !contextSet)
```

If you think that would be more understandable.

File: lib/circleci/token_validator_test.go (hunk 4 above)
Comment (strideynet) [reply lib/circleci/token_validator_test.go:91]: I was already doing this as part of the main test, but I'll move it into the helper for readabiltiy.

File: lib/githubactions/token_validator_test.go (hunk 5 above)
Comment (strideynet) [reply lib/githubactions/token_validator_test.go:47]: a8c047cd6f

File: lib/circleci/token_validator_test.go (hunk 4 above)
Comment (strideynet) [reply lib/circleci/token_validator_test.go:87]: a8c047cd6f

File: lib/auth/join_circleci.go (hunk 3 above)
Comment (strideynet) [reply lib/auth/join_circleci.go]: bf35db857b

File: api/types/provisioning.go (hunk 1 above)
Comment (zmb3) [reply api/types/provisioning.go:440]: Yeah, that's a bit easier for me to reason about. (You're right by the way, I had confused myself on the first review)

Comment (strideynet) [reply api/types/provisioning.go]: d12f98df97

File: lib/circleci/circleci.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+// Useful references:
+// - https://circleci.com/docs/openid-connect-tokens/
+
+package circleci
+
+import "fmt"
+
+const IssuerURLTemplate = "https://oidc.circleci.com/org/%s"
+
+func issuerURL(template, organizationID string) string {
```
Comment (timothyb89) [inline lib/circleci/circleci.go:34]: nit: I may have missed another usage, but does `template` need to be a parameter? It seems we only ever use the constant value.

**Review by timothyb89 — APPROVED** [review]
Looks good!

File: lib/circleci/circleci.go (hunk 6 above)
Comment (strideynet) [reply lib/circleci/circleci.go:34]: Yeah it's used as part of the test to make requests hit the test server instead :)

**Pull request merged**